{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quantify editing for Census Editrons in HEK cells\n",
    "\n",
    "writes an excel file w/ each sample as a seperate tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os, glob\n",
    "import shutil\n",
    "import matplotlib as plt\n",
    "import numpy as np\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO\n",
    "import fuzzysearch\n",
    "import re\n",
    "from collections import Counter, OrderedDict\n",
    "from Bio.Seq import Seq\n",
    "import xlsxwriter\n",
    "import gzip\n",
    "import pathlib\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import demultiplexing_module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#EXAMPLE 4\n",
    "#step through a number of samples, pull out replicates and write them to new fastqs in that file's folder named ???_rep1 or ???_rep2 etc.\n",
    "\n",
    "##Load key of conditions and files\n",
    "samples = pd.read_excel(\"Demultiplexing_Sample_key_AGK_26.xlsx\")[[\"condition\",\n",
    "                                           \"run\",\n",
    "                                           \"file\",\n",
    "                                           \"rep\",\n",
    "                                           \"rep_bc\"]]\n",
    "\n",
    "#step though samples\n",
    "for i in samples.index:\n",
    "    demultiplexing_module.demultiplex_to_file(samples['run'][i],samples['file'][i],samples['rep'][i],samples['rep_bc'][i]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>info</th>\n",
       "      <th>plasmid</th>\n",
       "      <th>L_search_seq</th>\n",
       "      <th>R_search_seq</th>\n",
       "      <th>wt_seq</th>\n",
       "      <th>edit_seq</th>\n",
       "      <th>run</th>\n",
       "      <th>rep_bc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>file_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>msAGK_26_198_rep1</th>\n",
       "      <td>pAGK85_EMX1_R1</td>\n",
       "      <td>85</td>\n",
       "      <td>AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC</td>\n",
       "      <td>CACATCAACCGGTGGCGCATTGCCACGAAG</td>\n",
       "      <td>AGAAGAAGAAGGGCTCCCAT</td>\n",
       "      <td>AGAAGAAAAAGTTCTCCCAT</td>\n",
       "      <td>msAGK_26</td>\n",
       "      <td>TCCTTTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAGK_26_199_rep1</th>\n",
       "      <td>pAGK90_EMX1_R1</td>\n",
       "      <td>90</td>\n",
       "      <td>AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC</td>\n",
       "      <td>CACATCAACCGGTGGCGCATTGCCACGAAG</td>\n",
       "      <td>AGAAGAAGAAGGGCTCCCAT</td>\n",
       "      <td>AGAAGAAAAAGTTCTCCCAT</td>\n",
       "      <td>msAGK_26</td>\n",
       "      <td>TCCTTTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAGK_26_200_rep1</th>\n",
       "      <td>pAGK95_EMX1_all_R1</td>\n",
       "      <td>95</td>\n",
       "      <td>AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC</td>\n",
       "      <td>CACATCAACCGGTGGCGCATTGCCACGAAG</td>\n",
       "      <td>AGAAGAAGAAGGGCTCCCAT</td>\n",
       "      <td>AGAAGAAAAAGTTCTCCCAT</td>\n",
       "      <td>msAGK_26</td>\n",
       "      <td>TCCTTTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAGK_26_200_rep2</th>\n",
       "      <td>pAGK95_EMX1_all_R2</td>\n",
       "      <td>95</td>\n",
       "      <td>AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC</td>\n",
       "      <td>CACATCAACCGGTGGCGCATTGCCACGAAG</td>\n",
       "      <td>AGAAGAAGAAGGGCTCCCAT</td>\n",
       "      <td>AGAAGAAAAAGTTCTCCCAT</td>\n",
       "      <td>msAGK_26</td>\n",
       "      <td>CTGCCCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAGK_26_200_rep3</th>\n",
       "      <td>pAGK95_EMX1_all_R3</td>\n",
       "      <td>95</td>\n",
       "      <td>AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC</td>\n",
       "      <td>CACATCAACCGGTGGCGCATTGCCACGAAG</td>\n",
       "      <td>AGAAGAAGAAGGGCTCCCAT</td>\n",
       "      <td>AGAAGAAAAAGTTCTCCCAT</td>\n",
       "      <td>msAGK_26</td>\n",
       "      <td>CGGTGCAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 info  plasmid  \\\n",
       "file_name                                        \n",
       "msAGK_26_198_rep1      pAGK85_EMX1_R1       85   \n",
       "msAGK_26_199_rep1      pAGK90_EMX1_R1       90   \n",
       "msAGK_26_200_rep1  pAGK95_EMX1_all_R1       95   \n",
       "msAGK_26_200_rep2  pAGK95_EMX1_all_R2       95   \n",
       "msAGK_26_200_rep3  pAGK95_EMX1_all_R3       95   \n",
       "\n",
       "                                     L_search_seq  \\\n",
       "file_name                                           \n",
       "msAGK_26_198_rep1  AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC   \n",
       "msAGK_26_199_rep1  AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC   \n",
       "msAGK_26_200_rep1  AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC   \n",
       "msAGK_26_200_rep2  AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC   \n",
       "msAGK_26_200_rep3  AGCTGGAGGAGGAAGGGCCTGAGTCCGAGC   \n",
       "\n",
       "                                     R_search_seq                wt_seq  \\\n",
       "file_name                                                                 \n",
       "msAGK_26_198_rep1  CACATCAACCGGTGGCGCATTGCCACGAAG  AGAAGAAGAAGGGCTCCCAT   \n",
       "msAGK_26_199_rep1  CACATCAACCGGTGGCGCATTGCCACGAAG  AGAAGAAGAAGGGCTCCCAT   \n",
       "msAGK_26_200_rep1  CACATCAACCGGTGGCGCATTGCCACGAAG  AGAAGAAGAAGGGCTCCCAT   \n",
       "msAGK_26_200_rep2  CACATCAACCGGTGGCGCATTGCCACGAAG  AGAAGAAGAAGGGCTCCCAT   \n",
       "msAGK_26_200_rep3  CACATCAACCGGTGGCGCATTGCCACGAAG  AGAAGAAGAAGGGCTCCCAT   \n",
       "\n",
       "                               edit_seq       run    rep_bc  \n",
       "file_name                                                    \n",
       "msAGK_26_198_rep1  AGAAGAAAAAGTTCTCCCAT  msAGK_26  TCCTTTCA  \n",
       "msAGK_26_199_rep1  AGAAGAAAAAGTTCTCCCAT  msAGK_26  TCCTTTCA  \n",
       "msAGK_26_200_rep1  AGAAGAAAAAGTTCTCCCAT  msAGK_26  TCCTTTCA  \n",
       "msAGK_26_200_rep2  AGAAGAAAAAGTTCTCCCAT  msAGK_26  CTGCCCTT  \n",
       "msAGK_26_200_rep3  AGAAGAAAAAGTTCTCCCAT  msAGK_26  CGGTGCAT  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load retrondb from backup csv for plasmids and umis\n",
    "#rdb = pd.read_csv(retrondb)\n",
    "#load key of conditions and files\n",
    "samples = pd.read_excel(\"Individual_filekey.xlsx\")\n",
    "samples = samples.set_index(\"file_name\")\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'198'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.index[0].split(\"_\")[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_and_exact_match(sequence, samples, file_name):\n",
    "        inner_left_flank = fuzzysearch.find_near_matches(samples.loc[file_name,\"L_search_seq\"],\n",
    "                                                         sequence,max_l_dist=4)\n",
    "        inner_right_flank = fuzzysearch.find_near_matches(samples.loc[file_name,\"R_search_seq\"],\n",
    "                                                          sequence,max_l_dist=4)\n",
    "        if len(inner_left_flank) == 1 and len(inner_right_flank) == 1:\n",
    "            region = sequence[inner_left_flank[0].end:inner_right_flank[0].start]\n",
    "            wt_or_edit = {\"wt\": samples.loc[file_name, \"wt_seq\"],\n",
    "                          \"edit\": samples.loc[file_name,\"edit_seq\"]}\n",
    "            for possibility in wt_or_edit.keys():\n",
    "                middle_search_seq = wt_or_edit[possibility]\n",
    "                if fuzzysearch.find_near_matches(middle_search_seq,region,max_l_dist=0):\n",
    "                    #change max_l_dist to \"1\" if 10bp barcode insert\n",
    "                    return (possibility, region)\n",
    "            return ('undetermined_no_site_match', region)\n",
    "        else:\n",
    "            return ('undetermined_no_flanking_match', None)\n",
    "\n",
    "def find_file(file_name, run):\n",
    "    miseq_folder_names = os.listdir('C:\\BaseSpace')\n",
    "    miseq_folder_dict = {}\n",
    "    for folder in miseq_folder_names:\n",
    "        miseq_folder_dict[folder.split('-')[0]] = folder\n",
    "    subfolder_names = os.listdir('C:\\BaseSpace\\%s' % miseq_folder_dict[run])\n",
    "    sub_folder_dict = {}\n",
    "    for folder in subfolder_names:\n",
    "        try:\n",
    "            intermediate_folder_key = folder.split('_')[2]\n",
    "            sub_folder_dict[intermediate_folder_key.split(\"-\")[0]] = folder\n",
    "        except IndexError: continue\n",
    "    file_folder = 'C:\\BaseSpace\\%s\\%s' % (miseq_folder_dict[run], sub_folder_dict[file_name.split('_')[2]])\n",
    "    for file in Path(file_folder).rglob(file_name + \"*\"):\n",
    "        file_path = file\n",
    "        return file\n",
    "        #return file_path\n",
    "#     print(file.split('_')[2])\n",
    "#     print(sub_folder_dict)\n",
    "#     print(miseq_folder_dict[run])\n",
    "#     print (file_folder)\n",
    "\n",
    "def get_editing(file_name, samples, run):\n",
    "    #input is the fastq and the editing group\n",
    "    #output is a dictionary of precise editing rate by plasmid name and indel rate\n",
    "    filepath = find_file(file_name, run)\n",
    "    print(filepath)\n",
    "    outcomes_dict = {'undetermined_no_site_match': 0,\n",
    "                    'undetermined_no_flanking_match': 0,\n",
    "                    'wt': 0,\n",
    "                    'edit': 0}\n",
    "  \n",
    "    all_reads_str = []\n",
    "    for seq_record in SeqIO.parse(filepath, \"fastq\"): \n",
    "        all_reads_str.append(str(seq_record.seq))\n",
    "    read_counter = Counter(all_reads_str)\n",
    "#     print (Region_dict_specific)\n",
    "    for read in read_counter:  \n",
    "    \n",
    "#     #S_number = file_name.split(\"_\")[2]\n",
    "#     #fastq_reads = '%s_S%s_L001_R1_001.fastq.gz' %(file_name, S_number)\n",
    "# #     fastq_reads = 'S%s.fastq' %(file_name, S_number)\n",
    "# #     print(fastq_reads)\n",
    "#     all_reads_str = []\n",
    "#     #adding gzip function... can get rid of this now\n",
    "# #     with gzip.open(fastq_reads,\"rt\") as handle:\n",
    "#     for seq_record in SeqIO.parse(handle, \"fastq\"): \n",
    "#         all_reads_str.append(str(seq_record.seq))\n",
    "#     read_counter = Counter(all_reads_str)\n",
    "\n",
    "# #     print (Region_dict_specific)\n",
    "#     for read in read_counter:\n",
    "        # get result of extract_and_exact_match\n",
    "        # (either wt, edit, undetermined_no_site_match, or undetermined_no_flanking_match)\n",
    "        site_match_result = extract_and_exact_match(read, samples, file_name)[0]\n",
    "        # adding the number of reads to our counter for that result\n",
    "        outcomes_dict[site_match_result] = outcomes_dict[site_match_result] + read_counter[read]\n",
    "    return (outcomes_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "msAGK_26_198_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_198-ds.b0405afdef3e406c8120ababb36abc7f\\msAGK_26_198_rep1.fastq\n",
      "279516 sequences do not match region (out of 487800 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  7673          3.683912\n",
      "undetermined_no_site_match           77673         37.291871\n",
      "wt                                  122938         59.024217\n",
      "\n",
      "\n",
      "msAGK_26_199_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_199-ds.d9b5745cf508437fabcfd9bcdf50db4d\\msAGK_26_199_rep1.fastq\n",
      "291482 sequences do not match region (out of 454712 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5911          3.621271\n",
      "undetermined_no_site_match           83462         51.131532\n",
      "wt                                   73857         45.247197\n",
      "\n",
      "\n",
      "msAGK_26_200_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_200-ds.50e7e776c4f14b9481ee3b0e404ff44d\\msAGK_26_200_rep1.fastq\n",
      "56034 sequences do not match region (out of 348665 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                 11805          4.034091\n",
      "undetermined_no_site_match          153790         52.554241\n",
      "wt                                  127036         43.411669\n",
      "\n",
      "\n",
      "msAGK_26_200_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_200-ds.50e7e776c4f14b9481ee3b0e404ff44d\\msAGK_26_200_rep2.fastq\n",
      "110640 sequences do not match region (out of 162233 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  1478          2.864730\n",
      "undetermined_no_site_match           25989         50.373113\n",
      "wt                                   24126         46.762158\n",
      "\n",
      "\n",
      "msAGK_26_200_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_200-ds.50e7e776c4f14b9481ee3b0e404ff44d\\msAGK_26_200_rep3.fastq\n",
      "168895 sequences do not match region (out of 279697 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  3100          2.797783\n",
      "undetermined_no_site_match           58537         52.830274\n",
      "wt                                   49165         44.371943\n",
      "\n",
      "\n",
      "msAGK_26_203_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_203-ds.108da57cc5cf4ab19f01711202c0d64f\\msAGK_26_203_rep1.fastq\n",
      "261657 sequences do not match region (out of 487217 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                   347          0.153839\n",
      "undetermined_no_site_match          104841         46.480316\n",
      "wt                                  120372         53.365845\n",
      "\n",
      "\n",
      "msAGK_26_208_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_208-ds.353c5cd0b54a47d8ba6cca01a94edba5\\msAGK_26_208_rep1.fastq\n",
      "481141 sequences do not match region (out of 571200 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  4218          4.683596\n",
      "undetermined_no_site_match           12660         14.057451\n",
      "wt                                   73181         81.258952\n",
      "\n",
      "\n",
      "msAGK_26_209_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_209-ds.6bc19b6931504b5cbc5ef524128e527c\\msAGK_26_209_rep1.fastq\n",
      "22024 sequences do not match region (out of 185937 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                 15395          9.392178\n",
      "undetermined_no_site_match           17746         10.826475\n",
      "wt                                  130772         79.781347\n",
      "\n",
      "\n",
      "msAGK_26_210_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_210-ds.8896a3fd9ff94293b1b76ce42b433a6c\\msAGK_26_210_rep1.fastq\n",
      "49968 sequences do not match region (out of 270258 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5247          2.381860\n",
      "undetermined_no_site_match           12623          5.730174\n",
      "wt                                  202420         91.887966\n",
      "\n",
      "\n",
      "msAGK_26_210_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_210-ds.8896a3fd9ff94293b1b76ce42b433a6c\\msAGK_26_210_rep2.fastq\n",
      "70781 sequences do not match region (out of 224675 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                 11820          7.680611\n",
      "undetermined_no_site_match           21781         14.153248\n",
      "wt                                  120293         78.166140\n",
      "\n",
      "\n",
      "msAGK_26_210_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_210-ds.8896a3fd9ff94293b1b76ce42b433a6c\\msAGK_26_210_rep3.fastq\n",
      "37114 sequences do not match region (out of 271151 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                 17950          7.669727\n",
      "undetermined_no_site_match           57047         24.375206\n",
      "wt                                  159040         67.955067\n",
      "\n",
      "\n",
      "msAGK_26_213_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_213-ds.586bcadfa66d4741be967ac9fd57c5f8\\msAGK_26_213_rep1.fastq\n",
      "380929 sequences do not match region (out of 498367 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  4975          4.236278\n",
      "undetermined_no_site_match           57108         48.628212\n",
      "wt                                   55355         47.135510\n",
      "\n",
      "\n",
      "msAGK_26_214_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_214-ds.ca189ddf0c9f4ab0965eb0c388ea5231\\msAGK_26_214_rep1.fastq\n",
      "185804 sequences do not match region (out of 353604 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  3879          2.311681\n",
      "undetermined_no_site_match           55932         33.332539\n",
      "wt                                  107989         64.355781\n",
      "\n",
      "\n",
      "msAGK_26_215_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_215-ds.833d93dc2cea4ef2a77880210b0f4f18\\msAGK_26_215_rep1.fastq\n",
      "199327 sequences do not match region (out of 273804 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  2226          2.988842\n",
      "undetermined_no_site_match           27290         36.642185\n",
      "wt                                   44961         60.368973\n",
      "\n",
      "\n",
      "msAGK_26_215_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_215-ds.833d93dc2cea4ef2a77880210b0f4f18\\msAGK_26_215_rep2.fastq\n",
      "57157 sequences do not match region (out of 305798 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  7590          3.052594\n",
      "undetermined_no_site_match           94197         37.884741\n",
      "wt                                  146854         59.062665\n",
      "\n",
      "\n",
      "msAGK_26_215_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_215-ds.833d93dc2cea4ef2a77880210b0f4f18\\msAGK_26_215_rep3.fastq\n",
      "110582 sequences do not match region (out of 316790 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  6097          2.956723\n",
      "undetermined_no_site_match           69126         33.522463\n",
      "wt                                  130985         63.520814\n",
      "\n",
      "\n",
      "msAGK_26_216_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_216-ds.c6c26b7c8e054a37b96a3c32b89974f3\\msAGK_26_216_rep1.fastq\n",
      "7320 sequences do not match region (out of 152025 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  1186          0.819598\n",
      "undetermined_no_site_match           97887         67.645900\n",
      "wt                                   45632         31.534501\n",
      "\n",
      "\n",
      "msAGK_26_217_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_217-ds.2bad37a241544f89a417337efcb27f3a\\msAGK_26_217_rep1.fastq\n",
      "32489 sequences do not match region (out of 258715 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5129          2.267202\n",
      "undetermined_no_site_match          157799         69.752814\n",
      "wt                                   63298         27.979985\n",
      "\n",
      "\n",
      "msAGK_26_218_rep1\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_218-ds.c32914bb64b5409d94dc36053eca23b2\\msAGK_26_218_rep1.fastq\n",
      "20818 sequences do not match region (out of 360095 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  9043          2.665374\n",
      "undetermined_no_site_match          170167         50.155772\n",
      "wt                                  160067         47.178854\n",
      "\n",
      "\n",
      "msAGK_26_218_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_218-ds.c32914bb64b5409d94dc36053eca23b2\\msAGK_26_218_rep2.fastq\n",
      "18505 sequences do not match region (out of 207579 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5612          2.968150\n",
      "undetermined_no_site_match           96816         51.205348\n",
      "wt                                   86646         45.826502\n",
      "\n",
      "\n",
      "msAGK_26_218_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_218-ds.c32914bb64b5409d94dc36053eca23b2\\msAGK_26_218_rep3.fastq\n",
      "33568 sequences do not match region (out of 502878 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  8118          1.729773\n",
      "undetermined_no_site_match          226892         48.345869\n",
      "wt                                  234300         49.924357\n",
      "\n",
      "\n",
      "msAGK_26_229_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_229-ds.f0f9ba5b04004d97929ec8702cb9e989\\msAGK_26_229_rep2.fastq\n",
      "158277 sequences do not match region (out of 181510 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                   306          1.317092\n",
      "undetermined_no_site_match           13356         57.487195\n",
      "wt                                    9571         41.195713\n",
      "\n",
      "\n",
      "msAGK_26_230_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_230-ds.416218b7f5f84992b37b4e2a3691a3e7\\msAGK_26_230_rep2.fastq\n",
      "571047 sequences do not match region (out of 701930 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  1464          1.118556\n",
      "undetermined_no_site_match           80461         61.475516\n",
      "wt                                   48958         37.405927\n",
      "\n",
      "\n",
      "msAGK_26_233_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_233-ds.1faa4b2c6405403fb53f8924549bafc1\\msAGK_26_233_rep2.fastq\n",
      "962634 sequences do not match region (out of 979059 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                    53          0.322679\n",
      "undetermined_no_site_match            8965         54.581431\n",
      "wt                                    7407         45.095890\n",
      "\n",
      "\n",
      "msAGK_26_229_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_229-ds.f0f9ba5b04004d97929ec8702cb9e989\\msAGK_26_229_rep3.fastq\n",
      "1077212 sequences do not match region (out of 1136874 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                   944          1.582247\n",
      "undetermined_no_site_match           35281         59.134793\n",
      "wt                                   23437         39.282961\n",
      "\n",
      "\n",
      "msAGK_26_230_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_230-ds.416218b7f5f84992b37b4e2a3691a3e7\\msAGK_26_230_rep3.fastq\n",
      "832802 sequences do not match region (out of 901059 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                   904          1.324406\n",
      "undetermined_no_site_match           41166         60.310298\n",
      "wt                                   26187         38.365296\n",
      "\n",
      "\n",
      "msAGK_26_233_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_233-ds.1faa4b2c6405403fb53f8924549bafc1\\msAGK_26_233_rep3.fastq\n",
      "711303 sequences do not match region (out of 747614 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                    12          0.033048\n",
      "undetermined_no_site_match           19828         54.606042\n",
      "wt                                   16471         45.360910\n",
      "\n",
      "\n",
      "msAGK_26_235_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_235-ds.be02f4194e544df8b2562573859987ab\\msAGK_26_235_rep2.fastq\n",
      "380102 sequences do not match region (out of 473707 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  2748          2.935741\n",
      "undetermined_no_site_match           17469         18.662465\n",
      "wt                                   73388         78.401795\n",
      "\n",
      "\n",
      "msAGK_26_236_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_236-ds.35cf84df00a24be8af33a354008635f9\\msAGK_26_236_rep2.fastq\n",
      "145583 sequences do not match region (out of 231189 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  3278          3.829171\n",
      "undetermined_no_site_match           11977         13.990842\n",
      "wt                                   70351         82.179987\n",
      "\n",
      "\n",
      "msAGK_26_235_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_235-ds.be02f4194e544df8b2562573859987ab\\msAGK_26_235_rep3.fastq\n",
      "240867 sequences do not match region (out of 348424 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  3474          3.229915\n",
      "undetermined_no_site_match           17865         16.609798\n",
      "wt                                   86218         80.160287\n",
      "\n",
      "\n",
      "msAGK_26_236_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_236-ds.35cf84df00a24be8af33a354008635f9\\msAGK_26_236_rep3.fastq\n",
      "71730 sequences do not match region (out of 398905 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                 19821          6.058226\n",
      "undetermined_no_site_match           39308         12.014365\n",
      "wt                                  268046         81.927409\n",
      "\n",
      "\n",
      "msAGK_26_240_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_240-ds.5b7e7b1b1cc74fe5a715f7b9f8429fb2\\msAGK_26_240_rep2.fastq\n",
      "96731 sequences do not match region (out of 248213 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  2940          1.940825\n",
      "undetermined_no_site_match           61567         40.643113\n",
      "wt                                   86975         57.416063\n",
      "\n",
      "\n",
      "msAGK_26_241_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_241-ds.bfe329f8d8eb44418df3440c98f9b11c\\msAGK_26_241_rep2.fastq\n",
      "199678 sequences do not match region (out of 277254 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  2178          2.807569\n",
      "undetermined_no_site_match           27497         35.445241\n",
      "wt                                   47901         61.747190\n",
      "\n",
      "\n",
      "msAGK_26_240_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_240-ds.5b7e7b1b1cc74fe5a715f7b9f8429fb2\\msAGK_26_240_rep3.fastq\n",
      "393027 sequences do not match region (out of 678014 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5637          1.977985\n",
      "undetermined_no_site_match          117129         41.099769\n",
      "wt                                  162221         56.922246\n",
      "\n",
      "\n",
      "msAGK_26_241_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_241-ds.bfe329f8d8eb44418df3440c98f9b11c\\msAGK_26_241_rep3.fastq\n",
      "365751 sequences do not match region (out of 475363 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  2900          2.645696\n",
      "undetermined_no_site_match           35629         32.504653\n",
      "wt                                   71083         64.849651\n",
      "\n",
      "\n",
      "msAGK_26_243_rep2\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_243-ds.be7c26eb171b43168c21b2127ba06a0a\\msAGK_26_243_rep2.fastq\n",
      "16722 sequences do not match region (out of 365381 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  5745          1.647742\n",
      "undetermined_no_site_match          196445         56.343017\n",
      "wt                                  146469         42.009241\n",
      "\n",
      "\n",
      "msAGK_26_243_rep3\n",
      "C:\\BaseSpace\\msAGK_26\\msAGK_26_243-ds.be7c26eb171b43168c21b2127ba06a0a\\msAGK_26_243_rep3.fastq\n",
      "15587 sequences do not match region (out of 407230 total sequences)\n",
      "\n",
      "                            raw read count  percent of reads\n",
      "edit                                  6384          1.630056\n",
      "undetermined_no_site_match          206925         52.835107\n",
      "wt                                  178334         45.534837\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "writer = pd.ExcelWriter('Individual_Processed.xlsx', engine='xlsxwriter')\n",
    "for file_name in samples.index.values:\n",
    "    print(file_name)\n",
    "    run = \"\"\n",
    "    outcomes = get_editing(file_name, samples, samples.loc[file_name, \"run\"])\n",
    "  \n",
    "    no_region = outcomes['undetermined_no_flanking_match']\n",
    "    del outcomes['undetermined_no_flanking_match']\n",
    "    region_seqs = sum(outcomes.values())\n",
    "    print ('%s sequences do not match region (out of %s total sequences)' %(no_region,region_seqs+no_region) +'\\n')\n",
    "    outcomes_plus = {}  #add percent editing\n",
    "    for key in outcomes:\n",
    "        try:\n",
    "            outcomes_plus[key] = [outcomes[key],(float(outcomes[key])/region_seqs)*100]\n",
    "        except ZeroDivisionError:\n",
    "            print ('no data to analyze' + '\\n')\n",
    "            break\n",
    "    sample_data = pd.DataFrame.from_dict(outcomes_plus, orient='index',columns=['raw read count','percent of reads'])\n",
    "    sorted_data = sample_data.sort_index(ascending=True)\n",
    "    # sorted_data.index.name = samples.loc[i]['info']\n",
    "    sorted_data.to_excel(writer, sheet_name=file_name)\n",
    "    for column in sorted_data:\n",
    "        column_length = max(sorted_data[column].astype(str).map(len).max(), len(column))\n",
    "        col_idx = sorted_data.columns.get_loc(column)\n",
    "        writer.sheets[file_name].set_column(col_idx, col_idx, column_length)\n",
    "    print (sorted_data)\n",
    "    print ('\\n')\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "sns.set_style(\"ticks\")\n",
    "from matplotlib.colors import LogNorm\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.rcParams['svg.fonttype'] = 'none'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\asim.khan\\AppData\\Local\\Temp\\ipykernel_9312\\2459494118.py:20: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  data = pd.concat([data, pd.DataFrame([run_data_dict])], ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# integrate data --> it's in sheets idk why\n",
    "data = pd.DataFrame(columns=[\"run_id\", \"edit_raw_read_count\",\n",
    "                             \"undetermined_no_site_match_raw_read_count\",\n",
    "                             \"wt_raw_read_count\", \"edit_percent\",\n",
    "                             \"undetermined_no_site_match_percent\",\n",
    "                             \"wt_percent\"])\n",
    "\n",
    "lib_filepath = \"Individual_Processed.xlsx\"\n",
    "lib_dict = pd.read_excel(lib_filepath, sheet_name=None)\n",
    "for run_id in lib_dict.keys():\n",
    "    run_data = lib_dict[run_id]\n",
    "    run_data_dict = {\"run_id\": run_id,\n",
    "                     \"edit_raw_read_count\": run_data.loc[run_data[\"Unnamed: 0\"] == \"edit\", \"raw read count\"][0],\n",
    "                     \"undetermined_no_site_match_raw_read_count\": run_data.loc[run_data[\"Unnamed: 0\"] == \"undetermined_no_site_match\", \"raw read count\"].values[0],\n",
    "                     \"wt_raw_read_count\": run_data.loc[run_data[\"Unnamed: 0\"] == \"wt\", \"raw read count\"].values[0],\n",
    "                     \"edit_percent\": run_data.loc[run_data[\"Unnamed: 0\"] == \"edit\", \"percent of reads\"].values[0],\n",
    "                     \"wt_percent\": run_data.loc[run_data[\"Unnamed: 0\"] == \"wt\", \"percent of reads\"].values[0],\n",
    "                     \"undetermined_no_site_match_percent\": run_data.loc[run_data[\"Unnamed: 0\"] == \"undetermined_no_site_match\",\n",
    "                                                                        \"percent of reads\"].values[0]}\n",
    "    data = pd.concat([data, pd.DataFrame([run_data_dict])], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(\"Individual_Processed_Clean.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
